typing/oprint.ml

```diff
@@ -459,38 +459,65 @@ let out_sig_item = ref (fun _ -> failwith "Oprint.out_sig_item")
 let out_signature = ref (fun _ -> failwith "Oprint.out_signature")
 let out_type_extension = ref (fun _ -> failwith "Oprint.out_type_extension")
 
-let rec print_out_functor funct ppf =
-  function
-    Omty_functor (_, None, mty_res) ->
-      if funct then fprintf ppf "() %a" (print_out_functor true) mty_res
-      else fprintf ppf "functor@ () %a" (print_out_functor true) mty_res
-  | Omty_functor (name, Some mty_arg, mty_res) -> begin
-      match name, funct with
-      | "_", true ->
-          fprintf ppf "->@ %a ->@ %a"
-            print_simple_out_module_type mty_arg
-            (print_out_functor false) mty_res
-      | "_", false ->
+let rec print_out_module_type ppf mty =
+  print_out_functor ppf mty
+and print_out_functor ppf = function
+  | Omty_functor _ as t ->
+     (* For anonymous functor arguments, the logic to choose between
+        the long-form
+          functor (_ : S) -> ...
+        and the short-form
+          S -> ...
+        is as follows: if we are already printing long-form functor arguments,
+        we use the long form unless all remaining functor arguments can use
+        the short form. (Otherwise use the short form.)
+
+        For example,
+          functor (X : S1) (_ : S2) (Y : S3) (_ : S4) (_ : S5) -> sig end
+        will get printed as
+          functor (X : S1) (_ : S2) (Y : S3) -> S4 -> S5 -> sig end
+
+        but
+          functor (_ : S1) (_ : S2) (Y : S3) (_ : S4) (_ : S5) -> sig end
+        gets printed as
+          S1 -> S2 -> functor (Y : S3) -> S4 -> S5 -> sig end
+      *)
+     let rec print_functor ppf = function
+       | Omty_functor ("_", Some mty_arg, mty_res) ->
           fprintf ppf "%a ->@ %a"
             print_simple_out_module_type mty_arg
-            (print_out_functor false) mty_res
-      | name, true ->
-          fprintf ppf "(%s : %a)@ %a" name
-            print_out_module_type mty_arg (print_out_functor true) mty_res
-      | name, false ->
-            fprintf ppf "functor@ (%s : %a)@ %a" name
-              print_out_module_type mty_arg (print_out_functor true) mty_res
-    end
-  | m ->
-      if funct then fprintf ppf "->@ %a" print_out_module_type m
-      else print_out_module_type ppf m
-and print_out_module_type ppf = function
-  | Omty_functor _ as t -> fprintf ppf "@[<2>%a@]" (print_out_functor false) t
+            print_functor mty_res
+       | Omty_functor _ as non_anonymous_functor ->
+          let rec split_args acc = function
+            | Omty_functor (name, mty_arg, mty_res) as func_mty
+                 when not (only_anonymous_args func_mty) ->
```

Comment:
Do we want to keep the quadratic complexity here? Or should we have a pair split_named_args/split_anonymous_args rather the current split_args/only_anonymous_args?

Reply:
I pushed another commit that implements a linear algorithm. There are several ways to do it, but most of them give code that is hard to read, I think the one I proposed (different from yours) is reasonably readable.

Reply:
(If you want to use a `split_anonymous_args` function, you must decide at the end of its loop whether to push back those arguments into the main accumulator, or to drop them and revert to the first anonymous abstraction as `rest` value. This requires keeping a copy of the `acc` and `rest` values of the last `split_named_args` call around,  and I found it hard to name them clearly and explain what they are for.)

Reply:
I agree that your implementation is very readable.